### PR TITLE
refactor(main): move remaining entry related code

### PR DESCRIPTION
**Pull request by ReenigneArcher**
## Description

This PR moves remaining entry point related code out of main. I wasn't sure of a better way to separate this.

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [x] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I want maintainers to keep my branch updated

### CODE REVIEW

File: src/config.cpp [hunk 1]
```diff
@@ -15,11 +15,12 @@
 #include <boost/property_tree/ptree.hpp>
 
 #include "config.h"
+#include "entry_handler.h"
 #include "file_handler.h"
 #include "logging.h"
-#include "main.h"
 #include "nvhttp.h"
 #include "rtsp.h"
+#include "thread_pool.h"
```
Comment (ReenigneArcher) [inline src/config.cpp:23]: This was generally added to handle `sleep_for` calls, although I'm not sure it's entirely necessary.

File: src/entry_handler.cpp [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+      nvprefs_instance.unload();
+    }
+    return 0;
+  }
+}  // namespace restore_nvprefs_undo
+#endif
+
+namespace lifetime {
+  char **argv;
+  std::atomic_int desired_exit_code;
```
Comment (ReenigneArcher) [inline src/entry_handler.cpp:109]: `static` was removed from these, as it wouldn't compile.

File: src/main.cpp [hunk 3]
```diff
@@ -37,22 +36,13 @@
 extern "C" {
 #include <libavutil/log.h>
 #include <rs.h>
```
Comment (ReenigneArcher) [inline src/main.cpp:38]: I'm not sure if any other includes can be removed from here now.

File: src/entry_handler.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  int entry(const char *name, int argc, char *argv[]);
+}  // namespace help
+
+namespace version {
+  int entry(const char *name, int argc, char *argv[]);
+}  // namespace version
+
+namespace lifetime {
+  extern char **argv;
+  extern std::atomic_int desired_exit_code;
```
Comment (ReenigneArcher) [inline src/entry_handler.h:51]: These two definitions were added.

File: src/main.cpp [hunk 5]
```diff
@@ -381,26 +68,12 @@ on_signal(int sig, FN &&fn) {
...
 std::map<std::string_view, std::function<int(const char *name, int argc, char **argv)>> cmd_to_func {
-  { "creds"sv, gen_creds::entry },
-  { "help"sv, help::entry },
-  { "version"sv, version::entry },
+  { "creds"sv, args::creds },
+  { "help"sv, args::help },
+  { "version"sv, args::version },
 #ifdef _WIN32
-  { "restore-nvprefs-undo"sv, restore_nvprefs_undo::entry },
+  { "restore-nvprefs-undo"sv, args::restore_nvprefs_undo },
```
Comment (ReenigneArcher) [inline src/main.cpp:76]: Refactored CLI args into one namepsace.

**Comment by ReenigneArcher** [comment]
```txt
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\sunshine_tests.dir/objects.a(upnp.cpp.obj):upnp.cpp:(.rdata$.refptr._ZN4mail3manE[.refptr._ZN4mail3manE]+0x0): undefined reference to `mail::man'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\sunshine_tests.dir/objects.a(stream.cpp.obj):stream.cpp:(.rdata$.refptr.task_pool[.refptr.task_pool]+0x0): undefined reference to `task_pool'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/../../../../x86_64-w64-mingw32/bin/ld.exe: CMakeFiles\sunshine_tests.dir/objects.a(video.cpp.obj):video.cpp:(.rdata$.refptr.display_cursor[.refptr.display_cursor]+0x0): undefined reference to `display_cursor'
```

If possible, I'd like to define these outside of `main.cpp` as well. Any idea?

File: src/main.cpp (hunk 3 above)
Comment (cgutman) [reply src/main.cpp:39]: I think most of those includes can probably go away now

File: src/config.cpp (hunk 1 above)
Comment (cgutman) [reply src/config.cpp:23]: You can replace it with `std::this_thread::sleep_for()`

File: src/stream.cpp [hunk 6]
```diff
@@ -18,14 +18,16 @@ extern "C" {
 }
 
 #include "config.h"
+#include "entry_handler.h"
```
Comment (cgutman) [inline src/stream.cpp:21]: What entry handler functions are these (stream.cpp, rtsp.cpp, etc) using?

Comment (ReenigneArcher) [reply src/stream.cpp:21]: Let me double check... it might have been from before I created the `globals`.

Comment (ReenigneArcher) [reply src/stream.cpp:21]: Here's a table of the files I added `entry_handler` to. I'll remove it from the ones without a check.

| source file                    | mail               | nv_prefs_instance  | lifetime           | service_ctlr       | launch_ui          | launch_ui_with_path |
|--------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|---------------------|
| src/audio.cpp                  | :white_check_mark: |                    |                    |                    |                    |                     |
| src/config.cpp                 |                    |                    |                    | :white_check_mark: | :white_check_mark: |                     |
| src/confighttp.cpp             | :white_check_mark: |                    |                    |                    |                    |                     |
| src/input.cpp                  |                    |                    |                    |                    |                    |                     |
| src/nvhttp.cpp                 | :white_check_mark: |                    |                    |                    |                    |                     |
| src/platform/linux/audio.cpp   |                    |                    |                    |                    |                    |                     |
| src/platform/linux/misc.cpp    |                    |                    | :white_check_mark: |                    |                    |                     |
| src/platform/linux/x11grab.cpp |                    |                    |                    |                    |                    |                     |
| src/platform/macos/misc.mm     |                    |                    | :white_check_mark: |                    |                    |                     |
| src/platform/windows/input.cpp |                    |                    |                    |                    |                    |                     |
| src/platform/windows/misc.cpp  |                    | :white_check_mark: | :white_check_mark: |                    |                    |                     |
| src/rtsp.cpp                   | :white_check_mark: |                    |                    |                    |                    |                     |
| src/stream.cpp                 |                    |                    |                    |                    |                    |                     |
| src/system_tray.cpp            |                    |                    | :white_check_mark: |                    | :white_check_mark: | :white_check_mark:  |
| src/upnp.cpp                   | :white_check_mark: |                    |                    |                    |                    |                     |
| src/video.cpp                  |                    |                    |                    |                    |                    |                     |

`src/platform/linux/audio.cpp` gives the following errors without `entry_handler.h`:

- No type named 'thread' in namespace 'std'; did you mean 'boost::thread'?: 201
- No member named 'thread' in namespace 'std': 238
- Expected ';' after expression: 238
- Expected ';' after expression: 239

Probably a better fix than inclduing `entry_handler.h` exists.

Comment (ReenigneArcher) [reply src/stream.cpp:21]: Solved audio.cpp by including `<thread>`

File: src/main.cpp (hunk 3 above)
Comment (ReenigneArcher) [reply src/main.cpp:29]: I think I removed everything that could be.

I never included `#include <libavutil/log.h>` anywhere else, so not sure about that?

File: src/stream.cpp (hunk 6 above)
Comment (cgutman) [reply src/stream.cpp]: I think moving the whole `mail` namespace into `globals.cpp/h` would clean things up nicely. It's a process-wide communication mechanism, so I think it makes sense for it to live with other global state.

Comment (ReenigneArcher) [reply src/stream.cpp]: Sounds good. I'll work on that.

Comment (ReenigneArcher) [reply src/stream.cpp]: Done in https://github.com/LizardByte/Sunshine/pull/2127/commits/a71d3b68ab325989f3ab3ecab9b74f280d8b2c62

File: src/main.cpp (hunk 3 above)
Comment (cgutman) [reply src/main.cpp:29]: Hmm, that should be for the `av_log_*` functions used below, but it must be getting pulled in via some indirect dependency. It's probably a good idea to leave it there anyway.

File: src/stream.cpp (hunk 6 above)
Comment (ReenigneArcher) [reply src/stream.cpp]: I made a few more tweaks, but I think this is ready now.

- moved `nvprefs_instance` to `globals`
- added/updated some documentation blocks in `globals.cpp` and `entry_handler.cpp`
- changed the input validation of the `creds` arg to the just use the `help` function within the `args` namespace

**Review by cgutman — APPROVED** [review]

**Pull request merged**